Initial Cupti PM sampling support for gpu profiler

cashton-nvidia opened the pull request:
Adds perf metric sampling to the Cupti gpu tracer.

Adds a unit test which validates expected number of flops are sampled.

rengoog commented on .clang-format:
```diff
@@ -4,3 +4,4 @@ PointerBindsToType: true
 SortIncludes: Never
 AlignTrailingComments:
   Kind: Always
+ColumnLimit: 80
```
probably remove? not sure - will differ to others.

cashton-nvidia replied:
Yeah I didn't mean to catch this change in the CL.  Can revert it w/o any issues.  However if the linter will enforce 80 character limits it also probably doesn't hurt

rengoog reviewed (approved):
Thanks @cashton-nvidia ! 2 more nits: other than that LGTM.

cashton-nvidia committed:
Initial Cupti PM sampling support for gpu profiler

cashton-nvidia committed:
Merge branch 'openxla:main' into add-cupti-pm-sampling

cashton-nvidia committed:
Add test for driver version 12.6 for PM sampling support

cashton-nvidia committed:
Merge remote-tracking branch 'refs/remotes/upstream-fork/add-cupti-pm-sampling' into add-cupti-pm-sampling

cashton-nvidia committed:
Merge branch 'openxla:main' into add-cupti-pm-sampling

cashton-nvidia commented on xla/backends/profiler/gpu/cupti_tracer.cc:
Thanks.  Can you refer me to an appropriate level for VLOG for these levels of detail?

allanrenucci replied:
Here is the Google internal guidline:
> In the source code you can specify VLOG(<level>) where <level> is usually a number from 1 to 10. A higher number indicates a chattier message. This allows you to turn on some or all VLOG messages for a given file. Since you will usually turn on VLOG for a given file, you can use whatever leveling scheme that makes sense for it. A typical pattern is to have high level messages use 1 and dumps of large protos use 10.

cashton-nvidia replied:
Thanks - Will update and get this in soon.

cashton-nvidia replied:
Done in recent commit

allanrenucci commented on xla/backends/profiler/gpu/profile_with_cuda_kernels_test.cc:
```diff
@@ -0,0 +1,171 @@
+/* Copyright 2024 The OpenXLA Authors.
```
Please add missing #ifndef header guard.

allanrenucci commented on xla/backends/profiler/gpu/profile_with_cuda_kernels.h:
```diff
@@ -0,0 +1,29 @@
+/* Copyright 2024 The OpenXLA Authors.
```
Please add missing #ifndef header guard.

cashton-nvidia replied:
(Done in next commit, thanks)

cashton-nvidia commented on xla/backends/profiler/gpu/cupti_interface.h:
```diff
@@ -20,10 +20,22 @@ limitations under the License.
 #include "tsl/platform/macros.h"
 #include "tsl/platform/types.h"
 
+// CUPTI PM sampling headers added in 12.6
+// Note - bug fix for PM sampling added after 12.6.2
+#if CUDA_VERSION >= 12060
+#define CUPTI_PM_SAMPLING 1
+#else
+#define CUPTI_PM_SAMPLING 0
+#endif
```
Are #define macros banned as well?  Or just the #if test?

Do you have an example of how XLA currently handles versioning in external dependencies where APIs / external headers & symbols are possibly not present, and does this without macros?

If not, first thoughts -
I'd imagine most / all of the #if cases in the tracer can be replaced with unimplemented errors.

The interface though seems more problematic - the new headers need to be avoided.  Since the preprocessor runs first, I'm not aware of any tricks except #if to avoid those includes.

Beyond including the headers, there's also the issue of the wrapper itself.  I could use #if #else to have alternate wrappers for these new possibly missing functions, or I could probably write some weak symbol replacements for the missing symbols.  I *might* be able to get the compiler to completely eliminate the references to the missing symbols and not need this, but I think that would rely on the compiler optimizations more than we can guarantee.  All of those options are more complicated to maintain than what's there currently.

cashton-nvidia replied:
I've done what I can - couldn't completely remove #if in a couple of cases - I needed it to include the headers, and to define a couple of types before use, but everywhere else I used compiler elimination of dead code to avoid errors due to missing functions / types.   Can you take a look?

allanrenucci replied:
> Are #define macros banned as well? Or just the #if test?

Just #if tests.

> I needed it to include the headers, and to define a couple of types before use, but everywhere else I used compiler elimination of dead code to avoid errors due to missing functions / types. Can you take a look?

Right, guarding new headers can't be avoided. Overall LGTM.

cashton-nvidia committed:
Switch how CUPTI versioning is handled, remove #if whereever possible

cashton-nvidia committed:
Merge remote-tracking branch 'upstream-fork/add-cupti-pm-sampling' into add-cupti-pm-sampling

allanrenucci commented on xla/backends/profiler/gpu/cupti_tracer.cc:
```suggestion
      if (!dev->FillCounterDataImage(info).ok()) continue;
```

allanrenucci commented on xla/backends/profiler/gpu/cupti_tracer.cc:
```suggestion
      if (!dev->GetSampleCounts(info).ok()) continue;
```

allanrenucci commented on xla/backends/profiler/gpu/cupti_tracer.cc:
Please replace all `tsl::errors::` usages with `absl::` equivalents. E.g. `absl::InternalError`.

cashton-nvidia replied:
Done

allanrenucci commented on xla/backends/profiler/gpu/cupti_tracer.h:
```diff
@@ -48,6 +72,207 @@ struct CuptiTracerOptions {
+class PmSamplingDevice {
+  // Internal state
+  // Declared roughly in order of initialization
+  char const* chipName_ = nullptr;
```
std::string?

cashton-nvidia replied:
The CUPTI APIs use char *, so this avoids converting to / from strings.  This chip name is produced by CUPTI and then later passed to other APIs.  It's not used by other XLA / xprof code.  Do we still need to convert it?

allanrenucci replied:
I am asking because it looks there is a memory leak.

We have `chipName_ = strdup(p.pChipName)` which IIUC creates a copy of `p.pChipName`. Should the copy be be freed when `PmSamplingDevice` is destroyed? If `chipName_` is stored as a `std::string` it will be freed automatically when `PmSamplingDevice` is destroyed.

> This chip name is produced by CUPTI and then later passed to other APIs.

Note you can just call `std::string::data` to get back a `const char*` when calling other APIs.

cashton-nvidia replied:
Ahh, true.  The string size would be small, ~6 bytes, but would be lost in this case.  Switching seemed more complex but w/ it being so simple I hadn't considered the leak.

allanrenucci commented on xla/backends/profiler/gpu/profile_with_cuda_kernels_test.cc:
```diff
@@ -0,0 +1,158 @@
+
+    CuptiTracerCollectorOptions collector_options;
+    collector_options.num_gpus = CuptiTracer::NumGpus();
+    auto start_walltime_ns = absl::GetCurrentTimeNanos();
+    auto start_gputime_ns = CuptiTracer::GetTimestamp();
```
Please don't use auto here. It only saves a few characters.

https://google.github.io/styleguide/cppguide.html#Type_deduction

cashton-nvidia replied:
Done

allanrenucci commented on xla/backends/profiler/gpu/profile_with_cuda_kernels_test.cc:
```diff
@@ -0,0 +1,158 @@
+std::atomic_uint64_t atomic_total_fp64 = 0;
+std::atomic_uint64_t atomic_total_read = 0;
+std::atomic_uint64_t atomic_total_write = 0;
+bool skip_first = true;
```
Should this be std::atomic_bool?

cashton-nvidia commented on xla/backends/profiler/gpu/profile_with_cuda_kernels_test.cc:
```diff
@@ -0,0 +1,158 @@
+namespace xla {
+namespace profiler {
+namespace test {
+
+using xla::profiler::CuptiInterface;
+using xla::profiler::CuptiTracer;
+using xla::profiler::CuptiTracerCollectorOptions;
+using xla::profiler::CuptiTracerOptions;
+using xla::profiler::CuptiWrapper;
+
+// Needed to create different cupti tracer for each test cases.
+class TestableCuptiTracer : public CuptiTracer {
+ public:
+  explicit TestableCuptiTracer()
+      : CuptiTracer(new CuptiErrorManager(std::make_unique<CuptiWrapper>())) {}
+};
+
+
+namespace {
```
Thanks for the style guide link.  I'd looked through xla source for a coding conventions reference and didn't find anything, though some sub parts of it referenced other conventions.  (I'm surrpised pragma once is forbidden since every compiler I'm aware of supports it, but I know it's technically non-standard.)  Will make these changes.

cashton-nvidia commented on xla/backends/profiler/gpu/mock_cupti.h:
```diff
@@ -91,8 +91,113 @@ class MockCupti : public xla::profiler::CuptiInterface {
   MOCK_METHOD(CUptiResult, GetGraphExecId,
               (CUgraphExec graph_exec, uint32_t* graph_id), (override));
 
+#if CUPTI_PM_SAMPLING
```
I missed this - I'm surprised it worked.  Tested by building with CUPTI_PM_SAMPLING forced to both 0 and 1, and it worked.  Must not be any live dependencies on the mock code?  Fixed.

cashton-nvidia commented on xla/backends/profiler/gpu/cupti_tracer.cc:
I wasn't clear on what vlog resolved to - if it's equivalent to log(info), I'll switch all these to the simpler code where possible

allanrenucci replied:
Yes, it is equivalent to `LOG(INFO)`. I'd say you only want to use `if (VLOG_IS_ON(...))` to guard against evaluating an expensive statement. E.g.
```c++
if (VLOG_IS_ON(1)) {
  Foo foo = Expensive();
  LOG(INFO) << foo;
}
```
But you can often rewrite as:
```c++
VLOG(1) << Expensive();
```
which is equivalent.

cashton-nvidia replied:
Done in next commit

allanrenucci commented on xla/backends/profiler/gpu/cupti_tracer.h:
```diff
@@ -48,6 +72,207 @@ struct CuptiTracerOptions {
+// - Buffer creation
+// - Worker thread creation / control
+// Decoding of counter data is done in PmSamplingDecodeThread class
+class PmSamplingDevice {
```
Please follow the following class format: https://google.github.io/styleguide/cppguide.html#Class_Format

> The public section should be first, followed by the protected and finally the private section.

allanrenucci replied:
See also https://google.github.io/styleguide/cppguide.html#Declaration_Order.

allanrenucci commented on xla/backends/profiler/gpu/cupti_tracer.h:
```diff
@@ -48,6 +72,207 @@ struct CuptiTracerOptions {
+// Responsible for fetching PM sampling data from device and providing it to
+// handler or other external container
+class PmSamplingDecodeThread {
+  enum thd_state {
```
Enumerators  should be named like constants: https://google.github.io/styleguide/cppguide.html#Enumerator_Names

allanrenucci commented on xla/backends/profiler/gpu/cupti_tracer.h:
```diff
@@ -48,6 +72,207 @@ struct CuptiTracerOptions {
+  }
+
+  // Signal and test for state transitions
+  bool IsThdInitialized() { return current_thd_state_ == state_initialized_; }
```
Please mark non-mutating method as const. I.e. IsThdInitialized, ShouldThdEnable, IsThdEnabled, ShouldThdDisable, IsThdEnabled, IsThdDisabled, ShouldThdExit, IsThdExiting

allanrenucci commented on xla/backends/profiler/gpu/cupti_tracer.h:
```diff
@@ -48,6 +72,207 @@ struct CuptiTracerOptions {
+
+  // Signal and test for state transitions
+  bool IsThdInitialized() { return current_thd_state_ == state_initialized_; }
+  void ThdIsInitialized() { current_thd_state_ = state_initialized_; }
```
I find this function name and its siblings confusing. ThdIsInitialized sounds like a boolean predicate, not a method that mutates the class state.

allanrenucci commented on xla/backends/profiler/gpu/cupti_tracer.cc:
```diff
@@ -1184,22 +1235,742 @@ absl::Status CuptiTracer::DisableApiTracing() {
+    return;
+  }
+  // Provide some defaults for metrics and handler
+  if (config_->metrics.size() == 0) {
```
```suggestion
  if (config_->metrics.empty()) {
```

allanrenucci commented on xla/backends/profiler/gpu/cupti_tracer.h:
```diff
@@ -48,6 +72,207 @@ struct CuptiTracerOptions {
+  absl::Status UnInitializeProfilerAPIs();
+  absl::Status DestroyCounterDataImage();
+
+  std::vector<char const*> default_metrics_{
```
```suggestion
  std::vector<std::string> default_metrics_ = {
```

allanrenucci commented on xla/backends/profiler/gpu/cupti_tracer.h:
```diff
@@ -31,6 +32,29 @@ limitations under the License.
+  size_t num_populated = 0;
+  int device_id;
+  std::vector<SamplerRange> sampler_ranges;
+  std::vector<char const*> metrics;
```
```suggestion
  std::vector<std::string> metrics;
```
Let's use std::string here it makes it easier to reason about ownership.

allanrenucci commented on xla/backends/profiler/gpu/cupti_tracer.h:
```diff
@@ -48,6 +72,207 @@ struct CuptiTracerOptions {
+  // Whether to enable PM sampling
+  bool enable_pm_sampling = false;
+  // List of metrics to enable
+  std::vector<const char*> metrics{};
```
```suggestion
  std::vector<std::string> metrics;
```
Let's use std::string here it makes it easier to reason about ownership.

allanrenucci commented on xla/backends/profiler/gpu/cupti_tracer.h:
```diff
@@ -48,6 +72,207 @@ struct CuptiTracerOptions {
+  volatile thd_state current_thd_state_ = state_uninitialized_;
+    
+  // State thread should transition to (only set by main thread)
+  volatile thd_state nextThdState_ = state_initialized_;
```
Please consistently use snake cases for variable names.

allanrenucci commented on xla/backends/profiler/gpu/cupti_tracer.h:
```diff
@@ -48,6 +72,207 @@ struct CuptiTracerOptions {
+class PmSamplingDevice {
+  // Internal state
+  // Declared roughly in order of initialization
+  char const* chipName_ = nullptr;
```
Please consistently use snake cases for variable names.

allanrenucci commented on xla/backends/profiler/gpu/cupti_tracer.cc:
const auto&?

allanrenucci commented on xla/backends/profiler/gpu/cupti_tracer.h:
```diff
@@ -48,6 +72,207 @@ struct CuptiTracerOptions {
+  volatile thd_state nextThdState_ = state_initialized_;
+
+  // Thread handle
+  std::thread* thd_;
```
Looking at initialisation (i.e. `thd_ = new std::thread(ThdFunc, this)`), it looks a memory leak. Use `std::unique_ptr<std::thread>` instead?

allanrenucci commented on xla/backends/profiler/gpu/cupti_tracer.cc:
```diff
@@ -1184,22 +1235,742 @@ absl::Status CuptiTracer::DisableApiTracing() {
+// Test for device support for PM sampling
+absl::Status PmSamplingDevice::DeviceSupported() {
+  ERROR_IF_PM_SAMPLING_NOT_SUPPORTED();
+  CUdevice cuDevice;
```
Please consistently use snake cases for variable names.

allanrenucci commented on xla/backends/profiler/gpu/cupti_tracer.cc:
const auto&?

allanrenucci commented on xla/backends/profiler/gpu/cupti_tracer.cc:
Please consistently use snake cases for variable names.

allanrenucci commented on xla/backends/profiler/gpu/cupti_tracer.cc:
const auto&?

allanrenucci commented on xla/backends/profiler/gpu/cupti_tracer.cc:
```suggestion
    while (!thd->IsThdInitialized());
```

allanrenucci commented on xla/backends/profiler/gpu/cupti_tracer.cc:
Log the error status, other you can just write:

```suggestion
    if (!dev->StartSampling().ok()) {
```

allanrenucci commented on xla/backends/profiler/gpu/cupti_tracer.cc:
```suggestion
  for (const auto& thd : option_->pm_sampling_config->threads) {
    while (!thd->IsThdEnabled());
```

allanrenucci commented on xla/backends/profiler/gpu/cupti_tracer.cc:
```suggestion
  for (const auto& thd : option_->pm_sampling_config->threads) {
    while (!thd->IsThdDisabled());
```

allanrenucci commented on xla/backends/profiler/gpu/cupti_tracer.cc:
```suggestion
    while (!thd->IsThdExiting());
```

allanrenucci commented on xla/backends/profiler/gpu/cupti_tracer.cc:
Log the error status? Otherwise you just write:

```c++
if (!dev->GetSample(info.sampler_ranges[i], i).ok())
```

allanrenucci commented on xla/backends/profiler/gpu/cupti_tracer.cc:
Log the error status? Otherwise you just write:

```c++
if (!dev->InitializeCounterDataImage().ok())
```

allanrenucci commented on xla/backends/profiler/gpu/cupti_tracer.cc:
Log the error status, other you can just write:

```suggestion
    if (!dev-> StopSampling().ok()) {
```

allanrenucci reviewed (changes requested):
@cashton-nvidia did you forget to push the last version?

cashton-nvidia commented:
> @cashton-nvidia did you forget to push the last version?

No, now that I've got the style guide, I realized there was more cleanup to do.  I do have all the changes I closed fixed locally, and will review what you've just sent along with any other changes I see in the guide.  Thanks!

cashton-nvidia commented on xla/backends/profiler/gpu/cupti_tracer.cc:
```diff
@@ -87,27 +95,56 @@ inline void LogIfError(const absl::Status &status) {
   } while (false)
 
-size_t Bytes2D(const CUDA_MEMCPY2D *p) { return p->Height * p->WidthInBytes; }
+#define RETURN_IF_CUDA_ERROR(expr)                                          \
```
I'm not sure why Bytes2D is showing up at those lines in the diff, odd.  Locally it's below RETURN_IF_CUDA_DRIVER_ERROR, and I don't recall moving it.  It isn't my code, but is used later on.  I'll double check this is in the right spot when I get this new update resolved.

cashton-nvidia commented on xla/backends/profiler/gpu/cupti_tracer.cc:
```diff
@@ -87,27 +95,56 @@ inline void LogIfError(const absl::Status &status) {
+    }                                                                       \
+  } while (false)
+
+#define RETURN_IF_ABSL_ERROR(expr)          \
```
Yep!  And removed.

cashton-nvidia committed:
Update to google style guide, fix a few minor issues

cashton-nvidia committed:
Merge branch 'openxla:main' into add-cupti-pm-sampling

cashton-nvidia committed:
Add definitions for types when cupti pm sampling headers are not present

allanrenucci commented on xla/backends/profiler/gpu/cupti_tracer.h:
```diff
@@ -20,17 +20,74 @@ limitations under the License.
+  // Whether to enable PM sampling
+  bool enable_pm_sampling = false;
+  // List of metrics to enable
+  std::vector<const char*> metrics{};
```
```suggestion
  std::vector<std::string> metrics;
```

allanrenucci commented on xla/backends/profiler/gpu/profile_with_cuda_kernels_test.cc:
```diff
@@ -0,0 +1,160 @@
+    tracer.Disable();
+
+    EXPECT_EQ(vec.size(), kNumElements);
+    for (int i = 0; i < kNumElements; ++i) {
+        EXPECT_GT(vec[i], (0. - 0.001)) << "index: " << i;
+        EXPECT_LT(vec[i], (0. + 0.001)) << "index: " << i;
+    }
```
```c++
#include "testing/base/public/gmock.h"

using ::testing::Each;
using ::testing::DistanceFrom;
using ::testing::Lt;

EXPECT_THAT(vec, Each(DistanceFrom(0, Lt(0.001))));
```

cashton-nvidia commented on xla/backends/profiler/gpu/cupti_error_manager.cc:
```diff
@@ -287,6 +287,373 @@ CUptiResult CuptiErrorManager::SetThreadIdType(
 
+// Profiler Host APIs
+CUptiResult CuptiErrorManager::ProfilerHostInitialize(
+    CUpti_Profiler_Host_Initialize_Params* pParams) {
```
Ahh, yes, since this was a 1:1 mapping to the CUPTI library I left the argument the same form as the wrapped func.  I can rename to whatever, I suppose.  BTW internally at least I can't get this to cleanly build w/ 12.3.2 yet so there will be an update here - might have to use templates to get constexpr to correctly eliminate the calls.

cashton-nvidia committed:
Build with new pm sampler interface, test hangs, commit to allow merge from main for updated env

cashton-nvidia committed:
Merge remote-tracking branch 'origin/main' into add-cupti-pm-sampling

cashton-nvidia committed:
Merge remote-tracking branch 'origin/main' into add-cupti-pm-sampling

cashton-nvidia committed:
Return error code from PM sampling device creation, skip test if perms error

cashton-nvidia committed:
Merge branch 'main' into add-cupti-pm-sampling

cashton-nvidia committed:
Update cupti_pm_sampler_impl.cc, fix typo

cashton-nvidia committed:
Fix clang-format issues

Most, not all.  Formatting demanded by clang-format for headers seems inconsistent (spaced pointer definition, etc)

cashton-nvidia committed:
Update cupti_pm_sampler_impl.cc

cashton-nvidia committed:
Change default parameters

cashton-nvidia committed:
Add validation of metrics and fallback to safe configurations when possible.

Simple algorithm to remove invalid metrics and shrink config until it fits in a single pass.

cashton-nvidia committed:
Merge remote-tracking branch 'origin/main' into add-cupti-pm-sampling

cashton-nvidia committed:
Switch to abseil synchronization for decode thread

cashton-nvidia committed:
Merge remote-tracking branch 'origin/main' into add-cupti-pm-sampling

cashton-nvidia committed:
Fix hlo_op profiler test, test back-to-back profiling fix

Lost the test for pm sampling actually being enabled in the options somewhere along the way, re-add that.
Also enables the possibly-problematic PmSamplingDisable call, which appears to fix back-to-back profiling (but may be terminating CUPTI too early)

cashton-nvidia committed:
Merge remote-tracking branch 'origin/main' into add-cupti-pm-sampling

cashton-nvidia committed:
Merge remote-tracking branch 'origin/main' into add-cupti-pm-sampling

cashton-nvidia committed:
More cleanup, removal of dead variables, style guide matching

cashton-nvidia committed:
Merge remote-tracking branch 'origin/main' into add-cupti-pm-sampling

cashton-nvidia commented:
Thanks for waiting for the update - there was a lot of backend work here getting additional testing runs through the pipeline.  I also refactored the code completely in order to hopefully make it build w/ CUDA versions prior to 12.6, which introduced the pm sampler code.  I believe it addresses all the comments here, but may have missed something.

rengoog commented on xla/backends/profiler/gpu/cupti_error_manager.h:
```diff
@@ -119,6 +119,103 @@ class CuptiErrorManager : public xla::profiler::CuptiInterface {
 
+  // Profiler Host APIs
+  CUptiResult ProfilerHostInitialize(
+      CUpti_Profiler_Host_Initialize_Params* pParams) override;
```
as @cashton-nvidia mentioned above, these are cuda driver symbols we have other examples of their use in these files as well.

allanrenucci commented:
@beckerhe are the usages of #ifdefs in this PR acceptable?

cashton-nvidia committed:
Merge remote-tracking branch 'origin/main' into add-cupti-pm-sampling

cashton-nvidia committed:
Update xla/backends/profiler/gpu/cupti_pm_sampler_impl.cc

Co-authored-by: Allan Renucci <allanrenucci@users.noreply.github.com>

cashton-nvidia commented on xla/backends/profiler/gpu/BUILD:
```diff
@@ -189,9 +190,43 @@ tsl_gpu_library(
+    ),
+)
+
+tsl_gpu_library(
```
I'll look at the definition.  I followed the pattern of cupti_tracer using this function, but I wasn't really paying attention to the need for it.

cashton-nvidia commented on xla/backends/profiler/gpu/BUILD:
```diff
@@ -455,3 +490,36 @@ xla_test(
+        "ptxas-options=-v",
+    ]),
+    tags = ["cuda-only"],
+    visibility = ["//visibility:public"],
```
Do you mean specifically the visibility flag or the test in general?

cashton-nvidia committed:
Gets to building with pm sampler factory.  Implements many changes from PR.  Currently failes due to metrics names re-arrangement.

cashton-nvidia committed:
Fixes string duplication issues, fixes bug between metrics names in device and thread classes

cashton-nvidia committed:
Small checkin to submit a few more PR review changes

cashton-nvidia committed:
Small commit for PR work before large change removing RETURN_IF_CUPTI_ERROR

cashton-nvidia committed:
Switch away from RETURN_IF_CUPTI_ERROR and other macros, enhance and share ToStatus

cashton-nvidia committed:
More PR cleanup before switching to weak syms for possibly missing funcs

cashton-nvidia committed:
Switch wrapper to using weak symbols instead of eliminated symbols

cashton-nvidia committed:
Revert "Switch wrapper to using weak symbols instead of eliminated symbols"

This reverts commit 4465edfefe7fc1e70c4a0f17fbe8707b59b40a79.

cashton-nvidia committed:
Reapply "Switch wrapper to using weak symbols instead of eliminated symbols"

This reverts commit 2445e8943226d1a37e2bb4637addc81c4dbda390.

cashton-nvidia committed:
Merge remote-tracking branch 'upstream-fork/add-cupti-pm-sampling' into add-cupti-pm-sampling

beckerhe reviewed (commented):
Hey, thank you for all the changes. I've added a few more comments, but in general the new changes look good. However some big points are still open, like the macro defines in cupti_interface.h.

cashton-nvidia committed:
Add missing absl call_once header

cashton-nvidia commented:
> Hey, thank you for all the changes. I've added a few more comments, but in general the new changes look good. However some big points are still open, like the macro defines in cupti_interface.h.

I'll push up fixes for all these points points before I do the merge from main, but I'm unsure of how to come to agreement on the macros in cupti_interface.h.  I don't see how to avoid including the headers w/o doing a preprocessor test, and I also couldn't figure out a way to skip the preprocessor block around the struct or function defintions (in cupti_interface.h and cupti_wrapper.cc).  At least, not with c++17.  I think we might be able to pull tricks w/ more modern c++ to get it to fully eliminate the missing definitions, but not here.

Do you have a solution for how to support building on older headers and not use these macros?

Thanks!

While waiting I will work through the merge from main.

cashton-nvidia committed:
Updates from PR review - move cupti status to separate unit, use cuda status from stream executor code, gpu device count bug fix, other minor fix up

cashton-nvidia committed:
Merge commit '4a5101096c' into add-cupti-pm-sampling

cashton-nvidia committed:
Merge commit '6fc3ab30e6' into add-cupti-pm-sampling

cashton-nvidia committed:
Merge commit '54d83f35e6' into add-cupti-pm-sampling

beckerhe commented:
> I'm working through merging main. I've pushed the merge that's been successful, up to ~June 15. [4a51010](https://github.com/openxla/xla/commit/4a5101096c6d9e71cb093b8f3dc44cc1ed2e1984). This seems OK, built and passed the unit test. Tried the next merge, ~10 days forwards to 6/25 [7af99c0](https://github.com/openxla/xla/commit/7af99c0463e2f61eaa1acfeee3a803a91048437a), and I hit missing cupti header definitions in bazel, and I am stuck. No matter what I do, bazel is missing the new 12.6 headers (cupti_pmsampling.h, for instance).
> 
> When I first tried to merge, with bazel just updating from the previous working build, it hit the missing definitions when building cupti_wrapper_stub.cc, but the headers seemed to be in all the right spots in bazel's output directories. I looked everywhere; could not get this working, so I did a full bazel expunge.
> 
> After the bazel expunge the behavior is different, now it fails during the actual compile of cupti_wrapper.cc (not the bazel analysis, but actual compiler invocation), it has a missing header error. Looking for the headers, now I see that bazel is missing these headers in half the generated directories, namely bazel-out/k8-opt does not have the new (from 12.6) cupti_pmsampling.h or cupti_profiler_host.h files.
> 
> I haven't figured it out from here - I'm not sure what results in this hermetic setup.

Can you push what you have even if it's failing? Then I can have a look.

cashton-nvidia commented:
> Can you push what you have even if it's failing? Then I can have a look.

It takes a long time for many of these builds to get to the point of failure, so I still haven't 100% isolated this to a particular commit, but I've got it narrowed down to one day of commits.  I just pushed up the merge of 54d83f35e6 which was clean, and then I fail when I try to `git merge 97fb0ca748`.  Hopefully that's actionable - I can push up the broken merge but this should be enough to reproduce.  When I've built the merge of 54d83f35e6 and then try to do this next merge, I get:

ERROR: /opt/xla/xla/backends/profiler/gpu/BUILD:149:16: Compiling xla/backends/profiler/gpu/cupti_wrapper.cc failed: undeclared inclusion(s) in rule '//xla/backends/profiler/gpu:cupti_wrapper':                                                                                                                                                       this rule is missing dependency declarations for the following files included by 'xla/backends/profiler/gpu/cupti_wrapper.cc':                                                'bazel-out/k8-opt/bin/external/cuda_cupti/_virtual_includes/headers/third_party/gpus/cuda/extras/CUPTI/include/cupti_pmsampling.h'                                          'bazel-out/k8-opt/bin/external/cuda_cupti/_virtual_includes/headers/third_party/gpus/cuda/extras/CUPTI/include/cupti_profiler_host.h'

and I see these warnings which I don't think I previously saw (or they didn't come out because there was no error?):
xla/backends/profiler/gpu/cupti_wrapper.cc:224:7: warning: comparison of function 'cuptiProfilerHostInitialize' not equal to a null pointer is always true [-Wtautological-pointer-compare]                                                                                                                                                               224 |   if (cuptiProfilerHostInitialize != nullptr) {                                                                                                                           |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~    ~~~~~~~                                                                                                                        xla/backends/profiler/gpu/cupti_wrapper.cc:224:7: note: prefix with the address-of operator to silence this warning                                                           224 |   if (cuptiProfilerHostInitialize != nullptr) {                                                                                                                           |       ^                                                                                                                                                                   |       &


Do I need to maybe regenerate my xla config?
# cat xla_configure.bazelrc
build --action_env CLANG_COMPILER_PATH=/usr/lib/llvm-18/bin/clang
build --repo_env CC=/usr/lib/llvm-18/bin/clang
build --repo_env BAZEL_COMPILER=/usr/lib/llvm-18/bin/clang
build --linkopt --ld-path=/usr/bin/ld.lld
build --config cuda_clang
build --action_env CLANG_CUDA_COMPILER_PATH=/usr/lib/llvm-18/bin/clang
build:cuda --repo_env HERMETIC_CUDA_COMPUTE_CAPABILITIES=8.6
build --config nonccl
build --action_env LD_LIBRARY_PATH=/usr/local/cuda/compat/lib:/usr/local/nvidia/lib:/usr/local/nvidia/lib64
build --action_env PYTHON_BIN_PATH=/usr/bin/python
build --python_path /usr/bin/python
test --test_env LD_LIBRARY_PATH
test --test_size_filters small,medium
build --copt -Wno-sign-compare
build --copt -Wno-error=unused-command-line-argument
build --copt -Wno-gnu-offsetof-extensions
build --build_tag_filters -no_oss,-rocm-only,-sycl-only
build --test_tag_filters -no_oss,-rocm-only,-sycl-only
test --build_tag_filters -no_oss,-rocm-only,-sycl-only
test --test_tag_filters -no_oss,-rocm-only,-sycl-only

I haven't done a bazel clean --expunge yet, but I expect that I will see a slightly different failure, with the headers really being missing.  (It seems bazel does not remove the headers from its output directories when this change happens)

cashton-nvidia committed:
Merge remote-tracking branch 'origin/main' into add-cupti-pm-sampling

Significant cleanup - adds additional thread safety analysis flags, logs flushes of pm sampling buffer, disables warnings on tautological-pointer-compare so we can continue using weak symbols as requested.

cashton-nvidia commented:
> Thanks. I'm going to have a look tomorrow.

With some help I was able to merge the update from main; notably the CUPTI header configuration was removed from XLA, so for this code to build this other PR will need to be applied first - https://github.com/google-ml-infra/rules_ml_toolchain/pull/46

I tested with the unit test included in this PR but this needs more testing.

I also had to work around a new error related to the weak symbols that were added recently - tautological-pointer-compare now triggers the symbol test against null when the header does define the symbols as regular (not weak).  In order to keep the requested weak symbol code there I disabled the warning for that section of the file.  The only other solution would be to remove the weak symbols and return to testing another constant to determine whether to call the function or return unsupported in the cupti wrapper code.  I prefer this original solution, but wanted to keep the requested weak symbol code there for now.

cashton-nvidia committed:
Revert change to allow CUPTI GetResultString() to always run, it was breaking a unit test

Perhaps this should be re-implemented in the future with the test fixed

cashton-nvidia commented:
Ahh yes I'd forgotten about this change.  I'll revert it.  I'd found that GetResultString wasn't printing anything after a CUPTI error due to the IGNORE_CALL_IF_DISABLED line.  I think it's more useful to have GetResultString always work (CUPTI's get result string function isn't stateful w/ the rest of CUPTI, and can be called after an error), but it looks like it broke this unit test which expected it to not work at this point.  I reverted that change, it fixes this test, and now `//xla/backends/profiler/gpu:all` passes for me.

cashton-nvidia committed:
Merge remote-tracking branch 'origin/main' into add-cupti-pm-sampling

cashton-nvidia committed:
Merge remote-tracking branch 'upstream/main' into add-cupti-pm-sampling

cashton-nvidia commented:
Pulled the updated rules_ml_toolchain ref into this branch.  At this point should be 100% complete (unless some of these comments weren't resolved satisfactorily)

cashton-nvidia committed:
Update weak symbols to remove for all but pm sampling and profiler target host

Allows intended use of weak symbols

cashton-nvidia commented:
Ahh, also found our discrepancy on the weak symbols - I'd missed that the tautological error was only for the always-existing functions when the weak symbols are in the exterrn C block.  Removed the symbol test for all but pm sampling and profiler host funcs, and it works as intended.  Updated.

beckerhe commented:
Hi Chris,
that's great. I think this is good to be merged.

I'm going to merge my version of your PR #29331 because it contains a bunch of fixes that are needed for this to work internally.
